parser/src/parse.rs

```diff
@@ -47,6 +47,98 @@ pub enum ParseUnicodeError {
+                                        string: s.to_string(),
+                                    })?
+                                    .1,
+                                chars
+                                    .next()
+                                    .ok_or(ParseError::InvalidEscape {
+                                        escape: "\\x".to_string(),
+                                        index: idx,
+                                        string: s.to_string(),
+                                    })?
+                                    .1,
+                            ]
+                            .iter()
+                            .collect();
+                            u8::from_str_radix(&hex, 16).map_err(|_| ParseError::InvalidEscape {
+                                escape: hex,
+                                index: idx,
+                                string: s.to_string(),
+                            })?
+                        }
+                        n if ('0'..='3').contains(&n) => {
+                            let octal: String = [
+                                n,
+                                chars
+                                    .next()
+                                    .ok_or(ParseError::InvalidEscape {
+                                        escape: format!("\\{n}"),
+                                        index: idx,
+                                        string: s.to_string(),
+                                    })?
+                                    .1,
+                                chars
+                                    .next()
+                                    .ok_or(ParseError::InvalidEscape {
+                                        escape: format!("\\{n}"),
+                                        index: idx,
+                                        string: s.to_string(),
+                                    })?
+                                    .1,
+                            ]
+                            .iter()
+                            .collect();
+                            u8::from_str_radix(&octal, 8).map_err(|_| {
+                                ParseError::InvalidEscape {
+                                    escape: octal,
+                                    index: idx,
+                                    string: s.to_string(),
+                                }
+                            })?
+                        }
+                        _ => {
+                            return Err(ParseError::InvalidEscape {
+                                escape: format!("{}{}", c, c2),
+                                index: idx,
+                                string: String::from(s),
+                            });
+                        }
+                    };
+
+                    res.push(byte);
+                    continue;
+                }
+            };
+        }
+        res.extend(c.to_string().as_bytes());
```

Comment:
Could this be simplified?
```suggestion
        res.push(c as u8);
```

Reply:
Ah! No we can't... or at least not _that_ way. I have not found a way to get to the bytes (plural) underlying a [`char`](https://doc.rust-lang.org/std/primitive.char.html) isn't a byte. It's actually even `sized` to 4 bytes, but holds _up to 4 bytes_.  

[Here an example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b526c8be5bd31b8f46932573bddbe7e0) that hopefully showcases the issue.

Reply:
Fascinating... you're right. TIL!

TL;DR, I'm approving and merging because I'm perfectly fine with what you have here.

But, if you're curious, I did some digging around and found you actually can do this
```rust
let required = c.len_utf8();
if res.capacity() - res.len() < required {
    res.reserve(required);
}
let mut buffer = [0; 4];
res.extend_from_slice(c.encode_utf8(&mut buffer).as_bytes());
```

After digging through the Rust source for `char` I saw that it does something very similar so I just had to benchmark it. Turns out, `encode_utf8` is 20x faster than the intermediate string conversion.
![image](https://github.com/user-attachments/assets/bb6df3c8-8629-4fdb-951b-0265607e96db)

```patch
Index: interpreter/benches/runtime.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/interpreter/benches/runtime.rs b/interpreter/benches/runtime.rs
--- a/interpreter/benches/runtime.rs	(revision b458dd53bdf2c169f58389e54201e9b38bec2659)
+++ b/interpreter/benches/runtime.rs	(date 1720826951595)
@@ -66,5 +66,62 @@
     group.finish();
 }
 
-criterion_group!(benches, criterion_benchmark, map_macro_benchmark);
+pub fn map_char_copy(c: &mut Criterion) {
+    let mut group = c.benchmark_group("char to vec");
+
+    group.bench_function("char to string to empty vec", |b| {
+        let mut empty = Vec::default();
+        let c = 'a';
+        b.iter(|| empty.extend(c.to_string().into_bytes()))
+    });
+
+    group.bench_function("char to string to full vec", |b| {
+        let mut full = Vec::with_capacity(4);
+        full.extend(vec![0, 0, 0, 0]);
+        assert_eq!(full.len(), 4);
+        assert_eq!(full.capacity(), 4);
+
+        let c = 'a';
+        b.iter(|| full.extend(c.to_string().into_bytes()))
+    });
+
+    group.bench_function("char encode to empty vec", |b| {
+        let mut empty = Vec::default();
+        let c = 'a';
+
+        b.iter(|| {
+            let required = c.len_utf8();
+            let available = empty.capacity() - empty.len();
+            if available < required {
+                empty.reserve(required);
+            }
+            let mut buffer = [0; 4];
+            c.encode_utf8(&mut buffer);
+            empty.extend_from_slice(&buffer[..required]);
+        })
+    });
+
+    group.bench_function("char encode to full vec", |b| {
+        let mut full = Vec::with_capacity(4);
+        full.extend(vec![0, 0, 0, 0]);
+        assert_eq!(full.len(), 4);
+        assert_eq!(full.capacity(), 4);
+        let c = 'a';
+
+        b.iter(|| {
+            let required = c.len_utf8();
+            let available = full.capacity() - full.len();
+            if available < required {
+                full.reserve(required);
+            }
+            let mut buffer = [0; 4];
+            c.encode_utf8(&mut buffer);
+            full.extend_from_slice(&buffer[..required]);
+        })
+    });
+
+    group.finish();
+}
+
+criterion_group!(benches, map_char_copy);
 criterion_main!(benches);
```

Reply:
Interesting! I can dig a little deeper in this. I did a quick run on [compiler explorer](https://rust.godbolt.org/) and the result didn't look outrageous, but indeed… it looks stupid (including when I wrote it), I was surprised there was no _obvious_ way to get the bytes. 
I might just steal your bench and adapt it to the `parse_byte` and alter it to `encode_utf8` to the buffer directly when no escaping is there… Good stuff! Thanks for looking into this!